core/src/replay_stage.rs

```diff
@@ -124,11 +124,40 @@ impl ReplayStage {
                         ticks_per_slot = bank.ticks_per_slot();
                     }
 
-                    // TODO: fork selection
-                    // vote on the latest one for now
-                    votable.sort_by(|b1, b2| b1.slot().cmp(&b2.slot()));
-
-                    if let Some(bank) = votable.last() {
+                    let locktower_start = Instant::now();
+                    // Locktower voting
+                    let flat_children = bank_forks.read().unwrap().flat_children();
+                    let flat_parents = bank_forks.read().unwrap().flat_parents();
+                    let frozen_banks = bank_forks.read().unwrap().frozen_banks();
+                    let mut votable: Vec<(u128, Arc<Bank>)> = frozen_banks
+                        .values()
+                        .filter(|b| b.is_votable())
+                        .filter(|b| !locktower.check_already_voted(b.slot()))
```

Comment:
why are these separate checks?

Reply:
@rob-solana keeping them separate makes it easier to test and change.